### PR TITLE
mo/password confirmation

### PR DESCRIPTION
**Pull request by MargaretOm**

### CODE REVIEW

File: app/javascript/components/ForgotPassword.jsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import Axios from 'axios';
+
+export default class ForgotPasswordForm extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { user: { email: '', redirect_url: 'http://localhost:3000/reset_password' } };
```
Comment (ermaac) [inline app/javascript/components/ForgotPassword.jsx]: do you think localhost will be working on staging?

File: app/javascript/components/ResetPassword.jsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+import Axios from 'axios';
+
+export default class ResetPasswordForm extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { user: { password: '',
+                           password_confirmation: '',
+                           "access-token": this.props.location.search.match(/access-token=(.+)&client=/)[1],
```
Comment (ermaac) [inline app/javascript/components/ResetPassword.jsx]: this parsing of params looks a little bit messy, can you parse with some js package?

File: app/javascript/components/SignIn.jsx [hunk 3]
```diff
@@ -16,18 +17,18 @@ export default class SigInForm extends React.Component {
       url: '/auth/sign_in',
       data: this.state.user
     }).then(response => {
-      sessionStorage.setItem('user',
+        sessionStorage.setItem('user',
```
Comment (ermaac) [inline app/javascript/components/SignIn.jsx]: I think alignment for this file was ok

File: app/javascript/components/SignUp.jsx [hunk 4]
```diff
@@ -16,18 +16,18 @@ export default class SignUpForm extends React.Component {
       url: '/auth',
       data: this.state.user
     }).then(response => {
-      sessionStorage.setItem('user',
+        sessionStorage.setItem('user',
```
Comment (ermaac) [inline app/javascript/components/SignUp.jsx]: I think alignment for this file was ok

File: config/environments/production.rb [hunk 5]
```diff
@@ -1,4 +1,5 @@
 Rails.application.configure do
+  config.action_mailer.default_url_options = { host: 'myherokuapp.herokuapp.com' }
```
Comment (ermaac) [inline config/environments/production.rb]: I think it should be taken from env variable

File: spec/requests/reset_password_spec.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  context 'incorrect data when forgot password' do
+    it 'return error message' do
+      post '/auth/password', params: { email: user.email }
+      expect(json_response['errors'].to_s).to include('Missing redirect URL')
+    end
+  end
+
+  context 'incorrect data when reset password' do
+    it 'return error message' do
+      put '/auth/password', params: { password: user.password, password_confirmation: user.password,
```
Comment (ermaac) [inline spec/requests/reset_password_spec.rb]: these params doesn't look like as incorrect data

File: spec/requests/reset_password_spec.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  it 'get reset password form' do
+    get '/reset_password'
+    expect(response).to have_http_status(:success)
+  end
+
+  context 'right params when forgot password' do
+    it 'return success' do
+      post '/auth/password', params: { email: user.email, redirect_url: 'http://localhost:3000/reset_password' }
```
Comment (ermaac) [inline spec/requests/reset_password_spec.rb]: Is email going to be sent after requesting this action? If so, please check that email was enqueued

File: spec/requests/reset_password_spec.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    get '/forgot_password'
+    expect(response).to have_http_status(:success)
+  end
+
+  it 'get reset password form' do
+    get '/reset_password'
+    expect(response).to have_http_status(:success)
+  end
+
+  context 'right params when forgot password' do
```
Comment (ermaac) [inline spec/requests/reset_password_spec.rb]: I think it is worth creating contexts for 'when forgot password', 'when reset password'

File: spec/requests/reset_password_spec.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  context 'incorrect data when forgot password' do
+    it 'return error message' do
+      post '/auth/password', params: { email: user.email }
+      expect(json_response['errors'].to_s).to include('Missing redirect URL')
+    end
+  end
+
+  context 'incorrect data when reset password' do
```
Comment (ermaac) [inline spec/requests/reset_password_spec.rb]: is anything prevent you from adding test for successful password reset?

File: app/javascript/components/ForgotPassword.jsx (hunk 1 above)
Comment (MargaretOm) [reply app/javascript/components/ForgotPassword.jsx]: I think on staging will be

File: app/javascript/components/ResetPassword.jsx [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  handleResetPassword = (e) => {
+    e.preventDefault();
+    Axios({
+      method: 'put',
+      url: '/auth/password',
+      data: this.state.user
+    }).then(response => {
+      this.props.history.push('/sign_in');
```
Comment (LuginaJulia) [inline app/javascript/components/ResetPassword.jsx]: Maybe after recovery password user must be authenticated and redirect to main page

File: config/environments/production.rb [hunk 11]
```diff
@@ -1,4 +1,6 @@
 Rails.application.configure do
+  Dotenv.load
+  config.action_mailer.default_url_options = { host: ENV['PRODUCTION_HOST' }
```
Comment (LuginaJulia) [inline config/environments/production.rb]: typo
```
{ host: ENV['PRODUCTION_HOST'] }
```

File: app/javascript/components/SignIn.jsx [hunk 12]
```diff
@@ -65,6 +66,8 @@ export default class SigInForm extends React.Component {
 
           <div>
             <label htmlFor="password">Password</label><br/>
+            <NavLink exact className="nav-link" activeClassName="active" to="/forgot_password"> (Forgot password?)</NavLink>
```
Comment (anatoliliotych) [inline app/javascript/components/SignIn.jsx]: maybe it's better to put it before buttons? :) or check some nice looking web app how they do this.

File: config/environments/production.rb [hunk 13]
```diff
@@ -1,4 +1,6 @@
 Rails.application.configure do
+  Dotenv.load
```
Comment (anatoliliotych) [inline config/environments/production.rb]: should be in application.rb

File: spec/requests/reset_password_spec.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    expect(response).to have_http_status(:success)
+  end
+
+  it 'get reset password form' do
+    get '/reset_password'
+    expect(response).to have_http_status(:success)
+  end
+
+  context 'right params when send letter' do
+    it 'return deliveries.size = 1' do
```
Comment (anatoliliotych) [inline spec/requests/reset_password_spec.rb]: too many ruby code in name of example.

File: spec/requests/reset_password_spec.rb [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+      post '/auth/password', params: { email: user.email, redirect_url: 'http://localhost:3000/reset_password' }
+      expect(ActionMailer::Base.deliveries.size).to eq(1)
```
Comment (anatoliliotych) [inline spec/requests/reset_password_spec.rb]: https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/change-matcher

File: spec/requests/reset_password_spec.rb [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    end
+  end
+
+  context 'incorrect data when send letter' do
+    it 'return error message' do
+      post '/auth/password', params: { email: user.email }
+      expect(ActionMailer::Base.deliveries.size).to eq(0)
```
Comment (anatoliliotych) [inline spec/requests/reset_password_spec.rb]: https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/change-matcher

File: spec/requests/reset_password_spec.rb [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    end
+  end
+
+  context 'right params when reset password' do
+    it 'return success message' do
+      get '/auth/password/edit', params: { reset_password_token: user.send(:set_reset_password_token),
+                                           redirect_url: 'http://localhost:3000/reset_password' }
+      access_token = response.headers['location'].match(/access-token=(.+)&client=/)[1]
+      client = response.headers['location'].match(/client=(.+)&client_id=/)[1]
+      put '/auth/password', params: { password: user.password, password_confirmation: user.password,
```
Comment (anatoliliotych) [inline spec/requests/reset_password_spec.rb]: follow style and be consistent in what you do

Comment (MargaretOm) [reply spec/requests/reset_password_spec.rb]: what are you meaning?

image

Comment (MargaretOm) [reply spec/requests/reset_password_spec.rb]: When I correct minor errors in the code, I write the name "Refactor code" and write a description in this commit, is this normal or no?
image

Comment (anatoliliotych) [reply spec/requests/reset_password_spec.rb]: be consistent means that if you make multiline hash - than all key/value pairs should be on separate lines.
About commits - the name Refactor code doesn't show what was done in commit. Commit name should answer on question Why the commit was done, and not how.

File: app/javascript/components/ForgotPassword.jsx [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    this.state = { user: { email: '', redirect_url: process.env.PRODUCTION_HOST + '/reset_password' } };
+    this.handleForgotPassword = this.handleForgotPassword.bind(this);
+    this.handleChange = this.handleChange.bind(this);
+  }
+
+  handleForgotPassword = (e) => {
+    delete this.state.response_message;
+    delete this.state.errors;
+    e.preventDefault();
+    debugger
```
Comment (ermaac) [inline app/javascript/components/ForgotPassword.jsx]: to be removed

File: app/javascript/components/ForgotPassword.jsx [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+import Axios from 'axios';
+import { FormButton } from '../ui/Buttons';
+import { FormTextField } from '../ui/TextFileds';
+import Grid from '@material-ui/core/Grid';
+require('dotenv').config();
+
+export default class ForgotPasswordForm extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { user: { email: '', redirect_url: process.env.PRODUCTION_HOST + '/reset_password' } };
```
Comment (ermaac) [inline app/javascript/components/ForgotPassword.jsx]: string interpolation

File: spec/requests/reset_password_spec.rb [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+    expect(response).to have_http_status(:success)
+  end
+
+  context 'right params when send letter' do
+    it 'return email delivery status plus one' do
+      expect {
+        post '/auth/password', params: { email: user.email,
+                                         redirect_url: 'http://localhost:3000/reset_password' }
+      }
+        .to(change { ActionMailer::Base.deliveries.size }.from(0).to(1))
```
Comment (ermaac) [inline spec/requests/reset_password_spec.rb]: by(1) better

File: spec/requests/reset_password_spec.rb [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+    end
+  end
+
+  context 'incorrect data when send letter' do
+    it 'return error message' do
+      expect {
+        post '/auth/password', params: { email: user.email }
+      }
+        .not_to(change { ActionMailer::Base.deliveries.size })
```
Comment (ermaac) [inline spec/requests/reset_password_spec.rb]: I'd put it on previous line

**Pull request merged**